Library/Homebrew/software_spec.rb

```diff
@@ -517,8 +517,11 @@ def compatible_locations?(tag: Utils::Bottles.tag)
 
     prefix = Pathname(cellar).parent.to_s
 
-    compatible_cellar = cellar == HOMEBREW_CELLAR.to_s
-    compatible_prefix = prefix == HOMEBREW_PREFIX.to_s
+    cellar_relocatable = cellar.size >= HOMEBREW_CELLAR.to_s.size && ENV["HOMEBREW_RELOCATE_BUILD_PREFIX"]
+    prefix_relocatable = prefix.size >= HOMEBREW_PREFIX.to_s.size && ENV["HOMEBREW_RELOCATE_BUILD_PREFIX"]
```

Comment:
Feels like more logic should be guarded by `HOMEBREW_RELOCATE_BUILD_PREFIX`? Do we want/need a way to turn this off for individual formulae?

Reply:
That's a good question.  Would it make sense to have a command line parameter for this (`--skip-prefix-relocation`), or formula DSL if we want to disable it at the formula level?

Reply:
I'm not sure, just posing the question. It seems likely that the rollout of something like this will look like:
- conditionally enabling this for known working formulae at bottle/CI time but do nothing at runtime for users
- same but enabled at runtime for all users
- conditionally enabling this for all formulae at bottle/CI time but do nothing at runtime for users
- enabled at runtime for all users

Reply:
The intention of this PR at the moment is to work with existing bottles, so there's not really anything to decide for bottle/CI time.  When part 2 of this (bottling with long prefixes) is ready, I think we'll need a new CLI parameter for `brew install` like `--long-prefix` which would be called by `brew test-bot`.  This would have to be explicitly enabled by a label in a PR or a parameter to a GitHub Action etc. on a per-formula basis.  I was even thinking we could have a new cellar type `:relocatable` which would indicate the bottle was built with a long prefix and can be relocated.

As for the rollout of this PR, if the user has a prefix that is too long, it just behaves as though binary patching doesn't exist.  If the user does have a short enough prefix, I think it should try to patch all non-relocatable formulae and then throw an error if it fails to fully relocate a formula telling the user to build that formula from source.  This wouldn't represent any sort of regression, just having to revert back to the old behavior.  

Of course we should first find out how many formulae fail to fully relocate after patching.  If it's a lot of them, we'd have to consider a formula DSL to explicitly enable/disable this so that installs/upgrades don't get interrupted by errors telling the user to build from source.

Reply:
> I was even thinking we could have a new cellar type :relocatable which would indicate the bottle was built with a long prefix and can be relocated.

I'd rather we re-used `:any` for this.

> If the user does have a short enough prefix, I think it should try to patch all non-relocatable formulae and then throw an error if it fails to fully relocate a formula telling the user to build that formula from source. This wouldn't represent any sort of regression, just having to revert back to the old behavior.

This still requires an environment variable to be set, right? As long as that behaviour is guarded 👍🏻

> Of course we should first find out how many formulae fail to fully relocate after patching. If it's a lot of them, we'd have to consider a formula DSL to explicitly enable/disable this so that installs/upgrades don't get interrupted by errors telling the user to build from source.

Formula DSL or just a list in test-bot/JSON in homebrew-core.

Reply:
> I'd rather we re-used `:any` for this.

In that case we'll need to add something to the bottle spec indicating that it needs to be relocated.  We're already planning to add a list of files that need to be relocated anyway, so the check could just be that this list exists.

> This still requires an environment variable to be set, right? As long as that behaviour is guarded 👍🏻

Yes, if the environment variable is not set, the formula is built from source like before.  The entire process is guarded by `HOMEBREW_RELOCATE_BUILD_PREFIX`.

> Formula DSL or just a list in test-bot/JSON in homebrew-core.

A list in homebrew-core would probably the simplest, though since the end user doesn't use `test-bot` to install formulae it will have be stored somewhere that makes sense for using with `brew install`/`brew upgrade` on their own.

Reply:
> In that case we'll need to add something to the bottle spec indicating that it needs to be relocated.

`:any` means it needs relocated. `:any_skip_relocation` means it does not. Are you thinking we need some other state for "only needs binary relocation and not other relocation" or something?

> Yes, if the environment variable is not set, the formula is built from source like before. The entire process is guarded by HOMEBREW_RELOCATE_BUILD_PREFIX.

Perfect 👍🏻 

> A list in homebrew-core would probably the simplest, though since the end user doesn't use test-bot to install formulae it will have be stored somewhere that makes sense for using with brew install/brew upgrade on their own.

I think in that case having something in `brew info --json` will be needed to ensure that it works with `HOMEBREW_INSTALL_FROM_API`.

Reply:
>`:any` means it needs relocated. `:any_skip_relocation` means it does not. Are you thinking we need some other state for "only needs binary relocation and not other relocation" or something?

No I think `:any` is fine for cases where binary relocation is the only thing needed.  What I was thinking about originally was making sure that `test-bot` knows a given bottle needs to be built with a long prefix so that binary relocation will work for all prefixes.  As long we store the information somewhere that `test-bot` can see before building a bottle, then we would only have to manually set the flag to use a long prefix the first time we want to make the bottle relocatable, and it would automatically be invoked in subsequent bottle runs.

>I think in that case having something in `brew info --json` will be needed to ensure that it works with `HOMEBREW_INSTALL_FROM_API`.

Making this work with `HOMEBREW_INSTALL_FROM_API` is very important so that's a good point about having it available in the JSON.  I have to familiarize myself more with the code for that but it makes sense to store the relocation info there.

Reply:
> As long we store the information somewhere that test-bot can see before building a bottle, then we would only have to manually set the flag to use a long prefix the first time we want to make the bottle relocatable, and it would automatically be invoked in subsequent bottle runs.

Gotcha. Yeh, some JSON in homebrew/core would make sense for me for this as it basically only affects our bottles.